diff --git a/ui/scene-view.reel/scene-view.js b/ui/scene-view.reel/scene-view.js
index d4e963f..1041b80 100755
--- a/ui/scene-view.reel/scene-view.js
+++ b/ui/scene-view.reel/scene-view.js
@@ -508,6 +508,16 @@ exports.SceneView = Component.specialize( {
         }
     },
 
+    // TSV: Use this to explicitly set 
+    //      this._viewPoint.glTFElement.transform.matrix on change
+    changeViewPointTrans: {
+        value: function(transMat /*mat4*/) {
+            if (this._viewPoint.glTFElement != null) {
+                this._viewPoint.glTFElement.transform.matrix = transMat;
+            }
+        }
+    },
+
     viewPoint: {
         get: function() {
             return this._viewPoint;
